### PR TITLE
feat: read-or-fetch for all pipelines

### PR DESCRIPTION
**Pull request by iloveagent57**
1. Refactors a common pattern for read-or-fetching a namedtuple of `CourseDetails` data from the DB, or from course-discovery if no record is present in the database.
2. Uses that pattern across all pipelines.

**Merge checklist:**
Check off if complete *or* not applicable:
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Fixup commits are squashed away
- [ ] Unit tests added/updated
- [ ] Manual testing instructions provided
- [ ] Noted any: Concerns, dependencies, migration issues, deadlines, tickets

### CODE REVIEW

File: federated_content_connector/filters/pipeline.py [hunk 1]
```diff
@@ -11,6 +12,48 @@
...
+    'id',
+    'course_key',
+    'course_type',
+    'product_source',
+    'start_date',
+    'end_date',
+    'enroll_by',
+]
+
+CourseDetailsData = namedtuple(
```
Comment (adamstankiewicz) [inline federated_content_connector/filters/pipeline.py:25]: Interesting! TIL about `namedtuple` :)

File: federated_content_connector/filters/pipeline.py [hunk 2]
```diff
@@ -35,36 +78,20 @@ def run_filter(self, course_key, course_home_url):  # pylint: disable=arguments-
...
-            return (None, None)
-
-        course_type = course_data.get('course_type')
-        product_source_value = course_data.get('product_source')
-        product_source = product_source_value
-        if isinstance(product_source_value, dict):
-            product_source = product_source_value['slug']
 
-        return course_type, product_source
+def is_exec_ed_2u_course(course_details_data):
```
Comment (adamstankiewicz) [inline federated_content_connector/filters/pipeline.py:90]: Nice helper method.

File: federated_content_connector/filters/pipeline.py [hunk 3]
```diff
@@ -11,6 +12,48 @@
...
+    """
+    model_instance = CourseDetails.objects.filter(id=course_key).first()
+    if model_instance:
+        return CourseDetailsData(**{
+            field_name: getattr(model_instance, field_name, None)
+            for field_name in COURSE_DETAILS_FIELDS
+        })
+
+    course_key_str = '{}+{}'.format(course_key.org, course_key.course)
+    course_data_dict = get_course_data(course_key_str, ['course_type', 'product_source'])
```
Comment (adamstankiewicz) [inline federated_content_connector/filters/pipeline.py:46]: [curious/sanity check] Does the `start_date` need to be passed to `get_course_data` to cover the `CreateApiRenderEnrollmentStep` pipeline's usage of `course_details.start_date`? I guess a similar question for `end_date` in `CreateApiRenderCourseRunStep`, too.

**Review by adamstankiewicz — COMMENTED** [review]
Looks good, just had one sanity check question before dropping a ✅

File: federated_content_connector/filters/pipeline.py (hunk 3 above)
Comment (iloveagent57) [reply federated_content_connector/filters/pipeline.py:46]: oh good call, gotta test this out better still

**Pull request closed without merging**